Use Rx5 by default

jseminck opened the pull request:
Started working on porting the library to use Rx5 by default.

* Updated the `getAdapter()` and test setup
* Changed peerDependency from `rx` to `rxjs`
* Updated README
* Updated blog example: changed to import local rxConnect instead of the version currently on NPM, and refactored the code to use Rx5 syntax
* Updated wikistream example: changed to import local rxConnect and removed setting up the adapter

What's left to do:

* Wikistream example is broken (as discussed). It seems wikipedia changed the API
* Update the docs + codepen examples. I don't think I can do this because they refer to the version hosted on NPM which is still the latest version that uses Rx4 by default. Once the new version is released, I can change these.
* I have a small wish to replace Codepen by Codesandbox ( https://codesandbox.io ) - it allows specifying the dependencies and pulls them straight from NPM, so it's easier to see what dependencies a project has.

jseminck committed:
Update blog example

jseminck committed:
Switch to using rx5 as default

jseminck committed:
Update readme to use Rx5 example. Remove the link to codepen that wasn't being shown

jseminck committed:
Update docs to explain how to use Rx4

jseminck committed:
Update wikistream example

jseminck commented:
BTW I am not sure if the adapter approach is really good. We found out that we were including both Rx4 and Rx5 in our bundle because currently rx-connect has a `peerDependency` on Rx4, even though we wanted to use Rx5. 

It turns out we had Rx4 in our `node_modules` directory because it was a dependency of one of our devDependencies (chimp) and webpack decided to bundle it anyway...

Anyway, this PR will fix that for us. But it might still impact users that want to use the Rx4 version... IMHO it's better just to advice them to use the older version of rx-connect instead (until some new features are added, if ever).

bsideup commented:
Hi @jseminck,

Great job! And thanks for pointing me at CodeSandbox, it looks great indeed :)

I'll move code examples from docs to the repo and point CodeSandbox at them, most probably today, so that we can fix the docs together with this PR, ok?

bsideup commented:
@jseminck done :) Now you can update the docs with RxJS5, see https://github.com/bsideup/rx-connect/tree/master/examples/docs for sources (FYI you can run locally with `yarn serve`)

bsideup commented on README.md:
```diff
@@ -53,7 +53,7 @@ with this:
 import { rxConnect } from "rx-connect";
 
 @rxConnect(
-    Rx.Observable.timer(0, 1000).timestamp()
```
I'm quite sure it's still there :)
http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-timestamp

jseminck replied:
That's weird. The RxJS4 to RxJS5 migration guide actually covers this one specifically:

https://github.com/ReactiveX/rxjs/blob/master/MIGRATION.md

<img width="694" alt="screen shot 2017-09-05 at 10 56 28" src="https://user-images.githubusercontent.com/9586897/30050980-0789fd54-9229-11e7-97d8-9772cff50f33.png">

I remember it not working at some point, but perhaps it got re-added? Will try because it makes the example a lot simpler and adds "wow factor" 😄

jseminck replied:
It seems it works but need to map it to a plain object now: https://codesandbox.io/s/vn9qjnqovl

bsideup replied:
@jseminck Hm... I'm thinking maybe we can relax the check here:
https://github.com/bsideup/rx-connect/blob/e586b5e31750a551387f5b700db7ed7ce22e2dd7/src/rxConnect.js#L67-L78

i.e. check if it's function first, if not - do the assignment. WDYT?

jseminck replied:
Yeah, will try it 😄

jseminck replied:
It doesn't really fix the issue here. `.timestamp()` returns an object, just not a plain object. From the docs: Observable<Timestamp<any>>

---

What seems to work is adding this (along with lodash.isobject). But I don't know what other implications it has ...


```js
if (isPlainObject(mutation)) {
    return Object.assign({}, state, mutation);
}

if (isObject(mutation)) {
    return Object.assign({}, state, {...mutation});
}
```

bsideup replied:
sounds good to me :)

jseminck replied:
I'll try it and run my branch on our code, to see if things still work 👍

bsideup commented on README.md:
```diff
@@ -81,3 +79,13 @@ class Timer extends React.PureComponent {
+
+## Using Rx4?
+
+This library supports Rx5 by default, but provides an adapter for Rx4:
```
library's actual name is "RxJS" :)

bsideup commented on examples/blog/src/actions/posts.js:
```diff
@@ -1,13 +1,14 @@
-            .getJSON(`//jsonplaceholder.typicode.com/posts?${ userId ? "userId=" + userId : ""}`)
+        return Rx.Observable
+            .ajax(`//jsonplaceholder.typicode.com/posts?${ userId ? "userId=" + userId : ""}`)
+            .map(e => e.response)
```
`.pluck("response")`?

bsideup commented on examples/blog/src/view/LoginView.jsx:
```diff
@@ -1,8 +1,8 @@
 import { connect } from "react-redux";
 import wrapActionCreators from "react-redux/lib/utils/wrapActionCreators";
-import { rxConnect, ofActions } from "rx-connect";
+import { rxConnect, ofActions } from "../../../../src";
```
Check Webpack's config and `alias` section. `from "rx-connect"` was aliasing to the sources actually

jseminck replied:
Ah cool. Yeah that's much nicer. I was confused because `rx-connect` is also specified as a dependency in the `package.json`: https://github.com/bsideup/rx-connect/blob/master/examples/blog/package.json#L19

bsideup commented on examples/blog/src/view/Time.jsx:
```diff
@@ -1,12 +1,12 @@
 export default class Time extends React.PureComponent {
     render() {
-        return <div>{new Date(this.props.timestamp).toLocaleTimeString()}</div>
+        return <div>{this.props.value} {new Date(this.props.timestamp).toLocaleTimeString()}</div>
```
`{this.props.value}` ?

jseminck replied:
Was testing stuff.. :D

bsideup commented on npm-debug.log:
```diff
@@ -0,0 +1,48 @@
+0 info it worked if it ends with ok
```
oopsy :D

jseminck committed:
Update Readme

jseminck committed:
Use pluck

jseminck committed:
Use webpack alias to rx-connect

I had to modify getAdapter() in rx-connect --- it was returning as an esModule probably because of the way it was being bundled
through webpack...

jseminck committed:
Return npm-debug.log

jseminck committed:
Update some docs

jseminck committed:
Remove console log

jseminck commented:
Tackled comments. Updated some examples. Also made examples on Code Sandbox with the current dependency but by changing the adapter: https://codesandbox.io/s/vn9qjnqovl

Some things...

**Timestamp**

Seems to work, but need to map to a plain object

```js
@rxConnect(
  Rx.Observable
    .timer(0, 1000)
    .timestamp()
    .map(timestamp => ({ ...timestamp }))
)
```

**JSONP**

It seems RxJS5 doesn't have JSONP support out of the box, which is needed for the Wikipedia examples. See https://github.com/ReactiveX/rxjs/issues/2095 & https://github.com/ReactiveX/rxjs/issues/1223

As a workaround:

```js
// RxJS5 doesn't support JSONP out of the box
function fetchFromWikipedia(search) {
  return fetchJsonp(
    `https://en.wikipedia.org/w/api.php?action=opensearch&search=${search}&format=json&callback=JSONPCallback`
  ).then(response => response.json());
}

function searchWikipedia(search) {
  return (
    Rx.Observable
      .from(fetchFromWikipedia(search))
      // Wikipedia has really weird response format o_O
      .map(([, titles, , urls]) =>
        titles.map((title, i) => ({ title, url: urls[i] }))
      )
  );
}
```

Maybe we can find a better example that doesn't need JSONP and we can use plain ajax?

**getAdapter()**

After changing the example to use webpack alias, I had to change `getAdapter()`. I think there's some issue in the build process for the blog example.

```js
export function getAdapter() {
    const adapter = rxConnect.adapter || require("./rx5Adapter");
    return adapter.__esModule ? adapter.default : adapter;
}
```

I'll go over the examples once more later, probably still missed some stuff... The main differences are:

* flatMapLatest to switchMap
* debounce to debounceTime
* .jsonp to ajax
* The timestamp thing

jseminck committed:
Merge branch 'master' into rx5-by-default

jseminck commented:
@bsideup have you seen my comment @ https://github.com/bsideup/rx-connect/pull/9#discussion_r137447317 ?

Otherwise, what else do we need to do to get this in? Would love to have this 😄

bsideup commented on examples/blog/src/actions/users.js:
```diff
@@ -1,13 +1,13 @@
 export function fetchUsers() {
     return () => {
-        return Rx.DOM.getJSON(`//jsonplaceholder.typicode.com/users`);
+        return Rx.Observable.ajax(`//jsonplaceholder.typicode.com/users`).map(e => e.response);
```
pluck? :)

bsideup commented:
@jseminck yeah, sorry for missing that. I left one comment re plain objects and minor one about pluck. Happy to merge once at least the first one is addressed :)

Commit:
Accept normal objects which are not arrays

Commit:
Pluck pluck pluck!! :)

bsideup commented:
@jseminck nice! Please also update the examples (Timers) and we're ready to go 🎉

jseminck commented:
Just pushed this now to run it with our project & see if everything remains
green. Will update examples during the day.



On Thu, 21 Sep 2017 at 09:29, Sergei Egorov <notifications@github.com>
wrote:

> @jseminck <https://github.com/jseminck> nice! Please also update the
> examples (Timers) and we're ready to go 🎉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bsideup/rx-connect/pull/9#issuecomment-331064151>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJJI0U5cXXL_FCMfOuoXceeCRgW8soVrks5skgIugaJpZM4PLJH3>
> .
>

Commit:
Update timestamp() examples

bsideup commented:
@jseminck merged! Thank a lot for your contribution :) Will release it in a few minutes

jseminck commented:
I will test it further, and update code sandbox to use the latest version
too!

On Mon, 25 Sep 2017 at 08:52, Sergei Egorov <notifications@github.com>
wrote:

> @jseminck <https://github.com/jseminck> merged! Thank a lot for your
> contribution :) Will release it in a few minutes
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bsideup/rx-connect/pull/9#issuecomment-331783484>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJJI0RBtDt-GljtxiXpNDUhJzbF5AGs9ks5slz-UgaJpZM4PLJH3>
> .
>

bsideup commented:
@jseminck released as 0.7.0 🎉 

I'll update CodeSandbox and remove JSONP :)

bsideup commented:
@jseminck done :) Thanks again :)
